### PR TITLE
Validate topics

### LINKED ISSUE #132: No topic checking - and crash
aedes allows using wildcards in PUBLISH:
* `base/#`
* `base/+/sub`

aedes allows using illegal wildcards in SUBSCRIBE and UNSUBSCRIBE:
* `base/#/sub` (multilevel wildcard not last element)
* `base/#sub` (wildcard not alone)
* `base/+xyz/sub` (wildcard not alone)

aedes does not check for empty (zero length) topics:
* in PUBLISH, aedes silently interrupts outgoing client connections (SUBSCRIBE to '#' in client A, send PUBLISH 'empty' from client B, connection to client A dies)
* in SUBSCRIBE aedes crashes
```
assert.js:60
  throw new errors.AssertionError({
  ^

AssertionError [ERR_ASSERTION]: '' == true
    at MQEmitter.on (/Users/ckrey/node_modules/mqemitter/mqemitter.js:74:3)
    at Aedes.subscribe (/Users/ckrey/node_modules/aedes/aedes.js:241:11)
    at SubscribeState.subTopic (/Users/ckrey/node_modules/aedes/lib/handlers/subscribe.js:106:12)
    at work (/Users/ckrey/node_modules/fastfall/fall.js:105:23)
    at /Users/ckrey/node_modules/aedes/lib/handlers/subscribe.js:74:5
    at MemoryPersistence.addSubscriptions (/Users/ckrey/node_modules/aedes-persistence/persistence.js:112:3)
    at SubscribeState.storeSubscriptions (/Users/ckrey/node_modules/aedes/lib/handlers/subscribe.js:73:9)
    at work (/Users/ckrey/node_modules/fastfall/fall.js:105:23)
    at Aedes.defaultAuthorizeSubscribe [as authorizeSubscribe] (/Users/ckrey/node_modules/aedes/aedes.js:306:3)
    at SubscribeState.authorize (/Users/ckrey/node_modules/aedes/lib/handlers/subscribe.js:39:17)
```

### PR DESCRIPTION
**Pull request by mcollina**
Fixes #132.

@ckrey is the correct behavior to severe the connection on a bad topic SUBSCRIBE?

### CODE REVIEW

**Comment by GavinDmello** [comment]
@mcollina We may also need to add handling for zero length topics

File: lib/handlers/publish.js [hunk 1]
```diff
@@ -17,6 +17,18 @@ var publishActions = [
   enqueuePublish
 ]
 function handlePublish (client, packet, done) {
+  var topic = packet.topic
+  var err
+  for (var i = 0; i < topic.length; i++) {
+    switch (topic.charCodeAt(i)) {
+      case 35:
+        err = new Error('# is not allowed in PUBLISH')
```
Comment (GavinDmello) [inline lib/handlers/publish.js]: We should probably disconnect the client when this behavior occurs

**Comment by ckrey** [comment]
No, I do understand the spec (section 4.7 Topic Names and Topic Filters)  that an invalid topic on PUBLISH or an invalid topic filter on SUBSCRIBE is a protocol violation and the broker should disconnect the client.

**Comment by ckrey** [comment]
@mcollina `+` isn't allowed in PUBLISH topics either

**Comment by mcollina** [comment]
@ckrey this should be handled by the PR, can you please check?

File: lib/handlers/publish.js (hunk 1 above)
Comment (mcollina) [inline lib/handlers/publish.js]: emitting `'error'`  would disconnect the client.

**Comment by mcollina** [comment]
@GavinDmello I added tests for the empty topics

**Comment by ckrey** [comment]
@mcollina sorry, I did not scroll to the end. Yes it is checked for `+` in PUBLISH topics.

But the checks for `+` and `#` in SUBSCRIBE topic filters are not complete (lib/handlers/subscribe.js):

`+` can occur at the end of a topic filter (e.g. `user/device/+`) 
`+` must be the only character on a level (e.g. `user/device+/abc` is not allowed)
`#` must be the only character on the last level (e.g. `user/device/xyz#` is not allowed)

**Comment by GavinDmello** [comment]
@mcollina Would this require changes in mqemitter for zero byte topics ?

File: lib/handlers/publish.js (hunk 1 above)
Comment (GavinDmello) [inline lib/handlers/publish.js]: 👍

**Comment by mcollina** [comment]
@ckrey @GavinDmello can you please check this is ok?

**Comment by ckrey** [comment]
Thanks, topic checks in PUBLISH and SUBSCRIBE are complete.

But, ... you may want to apply the topic filter checks to UNSUBSCRIBE too:

bildschirmfoto 2017-07-21 um 09 46 58
bildschirmfoto 2017-07-21 um 09 46 44
bildschirmfoto 2017-07-21 um 09 46 35
bildschirmfoto 2017-07-21 um 09 46 21

**Comment by mcollina** [comment]
what's the tool you are using?

**Comment by mcollina** [comment]
@ckrey updated.

**Comment by ckrey** [comment]
Checks are now ok for UNSUBSCRIBE too.

Not sure if you want to adapt the error message:

```
client error MQTTClient86134 # is only allowed in SUBSCRIBE in the last position Error: # is only allowed in SUBSCRIBE in the last position
...
client error MQTTClient67725 impossible to subscribe to an empty topic Error: impossible to subscribe to an empty topic
```

**Comment by ckrey** [comment]
@mcollina I am using the test cases my iOS MQTTClient library https://github.com/ckrey/MQTT-Client-Framework

Initially I used the tests to check functionality of the client library but it turned out to be quite a good tool for interoperability checks

**Comment by GavinDmello** [comment]
@ckrey Nice work on that repo 👍

**Comment by ckrey** [comment]
✅ (I will stop now 😉)

```
client error MQTTClient453651 # is only allowed in UNSUBSCRIBE in the last position Error: # is only allowed in UNSUBSCRIBE in the last position
client error MQTTClient853016 impossible to SUBSCRIBE to an empty topic Error: impossible to SUBSCRIBE to an empty topic
client error MQTTClient368645 # is not allowed in PUBLISH Error: # is not allowed in PUBLISH
```

**Comment by mcollina** [comment]
Thanks for the feedbacks and the reviews!

**Pull request merged**